packaging/build_rpm.py

```diff
@@ -0,0 +1,375 @@
+DEPENDENCIES_FILE = '{spec_file}.dependencies'
+RESULT_DIR = '/var/lib/mock/epel-7-x86_64/result'
+
+SCRIPT_DIR, SCRIPT_NAME = path_split(abspath(__file__))
+SSH_CONFIG_FILE = path_join(SCRIPT_DIR, 'ssh_config.tmp')
+
+RPMS_DIR = SCRIPT_DIR
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(SCRIPT_NAME)
+
+
+def main(args):
+    parser = argparse.ArgumentParser(
+            description="Build RPM using mock",
+            )
+    parser.add_argument(
+            'spec_file', type=argparse.FileType('r'),
+            help="which RPM to build",
+            )
+
+    args = parser.parse_args(args)
+
+    spec_file = abspath(args.spec_file.name)
+    packaging_dir, spec_file_name = path_split(spec_file)
+    # We assume that the spec always lives one dir level below the source
+    source = dirname(packaging_dir)
+
+    # Fetch any defined Sources (this does not include the repo containing the
+    # spec file itself, the local copy is to be used.
+    chdir(source)
+    sources = get_sources(spec_file)
+    sources += get_dependency_urls(spec_file)
+    logger.info(sources)
+    for url in sources:
+        download_if_newer(url)
+
+    chdir(SCRIPT_DIR)
+
+    if has_cmd('mock'):
+        # run locally
+        rpm = build(source, spec_file_name)
+        shutil.copy(rpm, SCRIPT_DIR)
+    elif has_cmd('vagrant'):
+        # install and run mock in a vagrant CentOS 7 box
+        check_call(['vagrant', 'up', 'builder'])
+
+        ssh_config = check_output(
+                ['vagrant', 'ssh-config', 'builder'])
+        with open(SSH_CONFIG_FILE, 'w') as f:
+            f.write(ssh_config)
+
+        install_mock()
+        # sync the code
+        for local, remote in (
+                (source, 'source'),
+                (SCRIPT_DIR + '/', 'build_script')):
+            check_call(
+                    ['rsync', '-avz',
+                     '-e', 'ssh -F {}'.format(SSH_CONFIG_FILE),
+                     '--exclude', '.git',
+                     '--exclude', '*.sw[op]',  # vim swap files
+                     '--delete', '--delete-excluded',
+                     local,
+                     'builder:' + remote,
+                     ])
+
+        source = path_join('source', basename(source))
+
+        run_vagrant(
+                'python "{this_script}" "{spec_file}"'.format(
+                    this_script=path_join('build_script', SCRIPT_NAME),
+                    spec_file=path_join(source, 'packaging', spec_file_name),
+                ))
+
+        # Copy files back
+        final_rpm = run_vagrant(
+            "find /var/lib/mock/epel-7-x86_64/result "
+            "-name '*.noarch.rpm' -o -name '*.x86_64.rpm'",
+            _func=check_output,
+            ).strip()
+        check_call(
+                ['scp', '-F', SSH_CONFIG_FILE,
+                 'builder:{rpm}'.format(rpm=final_rpm),
+                 '.'])
+
+        # Power down the vagrant box
+        check_call(['vagrant', 'halt', 'builder'])
```

Comment:
Maybe we need a big try except here, with a finally that halts the box? Just so we don't leave VMs running if something fails.

Reply:
During testing I've actually found it useful that the box is left up if the build fails, but I suppose always halting is probably better (there will only ever be one builder VM)